onnxruntime/core/providers/vitisai/imp/ep_context_utils.cc

```diff
@@ -0,0 +1,831 @@
+    p_node_attrs->emplace(kONNXModelFileNameAttr, *p_attr_3);
+    // Attr "partition_name".
+    auto p_attr_6 = ONNX_NAMESPACE::AttributeProto::Create();
+    p_attr_6->set_name(kPartitionNameAttr);
+    p_attr_6->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+    p_attr_6->set_s(fused_name);
+    p_node_attrs->emplace(kPartitionNameAttr, *p_attr_6);
+
+    p_ep_ctx_graph->AddNode(fused_name, kEPContextOp, "", input_node_arg_ptrs, output_node_arg_ptrs, p_node_attrs.get(), kEPContextOpDomain);
+
+    ++fused_index;
+  }
+  auto res_status = p_ep_ctx_graph->Resolve();
+  ORT_ENFORCE(res_status.IsOK(), res_status.ErrorMessage());
+  LOGS_DEFAULT(VERBOSE) << "Created EP context model graph resolved";
+}
+
+std::string RetrieveEPContextCache(
+    const Graph& graph, const PathString& ep_ctx_model_loc, bool binary_mode) {
+  // TODO: Support for multi-node EP context model.
+  const auto* p_node = GetEPContextNodePtr(graph);
+  const auto& attrs = p_node->GetAttributes();
+  int64_t embed_mode = attrs.at(kEmbedModeAttr).i();
+  const std::string& ep_ctx_cache = attrs.at(kEPCacheContextAttr).s();
+  if (embed_mode) {
+    return ep_ctx_cache;
+  }
+  fs::path ep_ctx_fs_path(ep_ctx_model_loc);
+  // Attr "ep_cache_context" stores a relative path.
+  ep_ctx_fs_path.replace_filename(fs::path(ep_ctx_cache));
+  LOGS_DEFAULT(VERBOSE) << "EP context cache path: " << ep_ctx_fs_path.string();
+  // TODO: Validaion of the file location to make sure security is met.
+  if (!fs::exists(ep_ctx_fs_path) || !fs::is_regular_file(ep_ctx_fs_path)) {
+    ORT_THROW("File for EP context cache is missing");
+  }
+  auto open_mode = binary_mode ? (std::ios::in | std::ios::binary) : std::ios::in;
+  std::ifstream ifs(ep_ctx_fs_path.string().c_str(), open_mode);
+  if (!ifs.is_open()) {
+    ORT_THROW("Exception opening EP context cache file");
+  }
+  ifs.seekg(0, ifs.end);
+  std::streampos cache_len = ifs.tellg();
+  if (cache_len == -1) {
+    ifs.close();
+    ORT_THROW("Error when operating EP context cache file");
+  } else if (cache_len == 0) {
+    ifs.close();
+    LOGS_DEFAULT(WARNING) << "Empty EP context cache file: " << ep_ctx_fs_path.string();
+    return "";
+  }
+  ifs.seekg(0, ifs.beg);
+  char* buf = new char[static_cast<size_t>(cache_len)];
+  ifs.read(buf, cache_len);
+  if (!ifs.good()) {
+    ifs.close();
+    ORT_THROW("Exception reading EP context cache file");
+  }
+  ifs.close();
+  std::string cache_payload(buf);
+  delete[] buf;
+  return cache_payload;
+}
+
+void RetrieveBackendCacheInfo(const Graph& graph, std::string& cache_dir, std::string& cache_key) {
+  // TODO: Support for multi-node EP context model.
+  const auto* p_node = GetEPContextNodePtr(graph);
+  const auto& attrs = p_node->GetAttributes();
+  const auto& notes_str = attrs.at(kNotesAttr).s();
+  nlohmann::json j_obj = nlohmann::json::parse(notes_str);
+  cache_dir = j_obj["backend_cache_dir"].get<std::string>();
```

Comment:
Be careful with the file path set inside the EPContext node. It could cause some security issue. Imagine that App can construct a model in memory, and create the ort session from that model memory. In that case it can set the dir to any place, so that it can access anything without user awareness.
That's why in Onnx if you save the model with external data (to avoid the proto buff 2GB limitation), the external data path only allow relative path inside the model folder. It doesn't allow path like ..\. Same for the ep_cache_context attribute in EPContext node. If embed_mode=0, it's the file path, and we only allow relative path, and don't allow path like ..\... There's code to check that. 
https://github.com/microsoft/onnxruntime/blob/4743803944d6f6333bf93b89a62d8083d4466710/onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc#L109-L129

Reply:
Thanks for such details.
We are also aware of this but haven't explicitly handled this potential security issue as fully as QNN EP yet.
For now, when "embed_mode" is "0", we just simply store a relative path in the attribute "ep_cache_context", but haven't covered different cases for security consideration yet.
https://github.com/glen-amd/onnxruntime/blob/227bbc3be94ed213be931a18f3059f8222475708/onnxruntime/core/providers/vitisai/imp/ep_context_utils.cc#L262-L263

Reply:
Anyway, I'll add logic to handle this in an iterative way.